rpc/src/module/chain.rs

```diff
@@ -364,4 +377,121 @@ impl ChainRpc for ChainRpcImpl {
+        let mut tx_indices = HashSet::new();
+        for tx_hash in tx_hashes {
+            match snapshot.get_transaction_info(&tx_hash.pack()) {
+                Some(tx_info) => {
+                    if retrieved_block_hash.is_none() {
+                        retrieved_block_hash = Some(tx_info.block_hash);
+                    } else if Some(tx_info.block_hash) != retrieved_block_hash {
+                        return Err(RPCError::invalid_params(
+                            "Not all transactions found in retrieved block",
+                        ));
+                    }
+
+                    if !tx_indices.insert(tx_info.index as u32) {
+                        return Err(RPCError::invalid_params(format!(
+                            "Duplicated tx_hash {:#x}",
+                            tx_hash
+                        )));
+                    }
+                }
+                None => {
+                    return Err(RPCError::invalid_params(format!(
+                        "Transaction {:#x} not yet in block",
+                        tx_hash
+                    )));
+                }
+            }
+        }
+
+        let retrieved_block_hash = retrieved_block_hash.expect("checked len");
+        if let Some(specified_block_hash) = block_hash {
+            if !retrieved_block_hash.eq(&specified_block_hash.pack()) {
+                return Err(RPCError::invalid_params(
+                    "Not all transactions found in specified block",
+                ));
+            }
+        }
+
+        snapshot
+            .get_block(&retrieved_block_hash)
+            .ok_or_else(|| {
+                let message = format!(
+                    "Chain TransactionInfo says block {:#x} existing, but that block is not in the database",
+                    retrieved_block_hash
+                );
+                error!("{}", message);
+                RPCError::custom(RPCError::ChainIndexIsInconsistent, message)
+            })
+            .map(|block| {
+                let proof = CBMT::build_merkle_proof(
+                    &block.transactions().iter().map(|tx| tx.hash()).collect::<Vec<_>>(),
+                    &tx_indices.into_iter().collect::<Vec<_>>())
+                .expect("build proof with verified inputs should be OK");
+                TransactionProof {
+                    block_hash: block.hash().unpack(),
+                    witnesses_root: block.calc_witnesses_root().unpack(),
+                    proof: JsonMerkleProof {
+                        indices: proof.indices().iter().map(|index| (*index).into()).collect(),
+                        lemmas: proof.lemmas().iter().map(|lemma| Unpack::<H256>::unpack(lemma)).collect(),
+                    }
+                }
+            })
+    }
+
+    fn verify_transaction_proof(&self, tx_proof: TransactionProof) -> Result<Vec<H256>> {
+        let snapshot = self.shared.snapshot();
+
+        snapshot
+            .get_block(&tx_proof.block_hash.pack())
```

Comment:
I'm not sure whether it is a performance concern to read the whole block to only verify several leaves. As I mentioned in https://github.com/nervosnetwork/merkle-tree/pull/20/files#r484305194 , its possible to only read the transactions in the proof.

Reply:
It may be a bottleneck when this RPC is called frequently, and the optimized code is able to directly read the tx hash stored inside `COLUMN_BLOCK_BODY`, but I don't see the need here to optimize early, it makes the code more complex.